File: ios/MyLiftSquad/Sources/Data/PrCalculator.swift
import Foundation

enum PrCalculator {

    private static let invalidPlaces: Set<String> = ["DQ", "DD", "DNS", "NS", "G"]

    struct Result {
        let bestSquat: Double
        let bestBench: Double
        let bestDeadlift: Double
        let bestTotal: Double
    }

    static func calculate(from results: [CompetitionResult]) -> Result {
        let valid = results.filter { !invalidPlaces.contains($0.place) }
        return Result(
            bestSquat: valid.map(\.best3SquatKg).filter { $0 > 0 }.max() ?? 0,
            bestBench: valid.map(\.best3BenchKg).filter { $0 > 0 }.max() ?? 0,
            bestDeadlift: valid.map(\.best3DeadliftKg).filter { $0 > 0 }.max() ?? 0,
            bestTotal: valid.map(\.totalKg).filter { $0 > 0 }.max() ?? 0
        )
    }
}

File: android/app/src/main/java/com/gooseco/myliftsquad/data/PrCalculator.kt
package com.gooseco.myliftsquad.data

import com.gooseco.myliftsquad.data.db.CompetitionEntry

object PrCalculator {

    private val disqualifiedPlaces = setOf("DQ", "DD", "DNS", "NS", "G")

    data class Result(
        val bestSquat: Double?,
        val bestBench: Double?,
        val bestDeadlift: Double?,
        val bestTotal: Double?
    )

    fun calculate(entries: List<CompetitionEntry>): Result {
        val valid = entries.filter { it.place !in disqualifiedPlaces }
        return Result(
            bestSquat = valid.mapNotNull { it.best3SquatKg?.takeIf { v -> v > 0 } }.maxOrNull(),
            bestBench = valid.mapNotNull { it.best3BenchKg?.takeIf { v -> v > 0 } }.maxOrNull(),
            bestDeadlift = valid.mapNotNull { it.best3DeadliftKg?.takeIf { v -> v > 0 } }.maxOrNull(),
            bestTotal = valid.mapNotNull { it.totalKg?.takeIf { v -> v > 0 } }.maxOrNull()
        )
    }
}
